src/samples/http/server-sse.c

```diff
@@ -0,0 +1,295 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @file
+ * @brief HTTP Server sse
+ *
+ * Sample server that return a response and keep it alive.  It
+ * implements server sent events. The server sends the data typed in
+ * stdin and the data is broadcasted to all the clients connected.
+ *
+ * To test it:
+ *
+ * run ./server-sse -p 8080
+ * open a browser in http://127.0.0.1:8080 or curl http://127.0.0.1:8080
+ * type something
+ *
+ * To see the usage help, -h or --help.
+ */
+
+
+#include <getopt.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <fcntl.h>
+
+#include "soletta.h"
+#include "sol-http.h"
+#include "sol-http-server.h"
+#include "sol-util.h"
+#include "sol-util-file.h"
+
+#define HTML_FILE \
+    "<!DOCTYPE html>" \
+    "<html>" \
+    "<body>" \
+    "<h1>Getting server updates</h1>" \
+    "<div id=\"result\"></div>" \
+    "<script>" \
+    "if(typeof(EventSource) !== \"undefined\") {" \
+    "var source = new EventSource(\"http://%.*s:%d/events\");" \
+    "source.onmessage = function(event) {" \
+    "document.getElementById(\"result\").innerHTML += event.data + \"<br>\";" \
+    "};" \
+    "} else {" \
+    "document.getElementById(\"result\").innerHTML = \"Sorry, your browser does not support server-sent events...\";" \
+    "}" \
+    "</script>" \
+    "</body>" \
+    "</html>"
+
+
+static struct sol_http_server *server;
+static struct sol_fd *stdin_watch;
+static struct sol_buffer value;
+static struct sol_ptr_vector responses = SOL_PTR_VECTOR_INIT;
+static int port = 8080;
+
+static bool
+on_stdin(void *data, int fd, uint32_t flags)
+{
+    uint16_t i;
+    struct sol_http_response_progressive *sse;
+    static const struct sol_str_slice begin = {
+        .len = sizeof("data: ") - 1,
+        .data = (const char *)"data: "
+    };
+    static const struct sol_str_slice end = {
+        .len = sizeof("\n\n") - 1,
+        .data = (const char *)"\n\n"
+    };
+
+    if (flags & (SOL_FD_FLAGS_ERR | SOL_FD_FLAGS_HUP)) {
+        fprintf(stderr, "ERROR: Something wrong happened with file descriptor: %d\n", fd);
+        goto err;
+    }
+
+    if (flags & SOL_FD_FLAGS_IN) {
+        int err;
+
+        value.used = 0;
+        /* this will loop trying to read as much data as possible to buffer. */
+        err = sol_util_load_file_fd_buffer(fd, &value);
+        if (err < 0) {
+            fprintf(stderr, "ERROR: failed to read from stdin: %s\n",
+                sol_util_strerrora(-err));
+            goto err;
+        }
+
+        if (value.used == 0) {
+            /* no data usually means ^D on the terminal, quit the application */
+            printf("no data on stdin, quitting.\n");
+            sol_quit();
```

Comment:
we should close the progressive connection and wait for it to gracefully exit before we do quit, otherwise we may lose data.